### DISCUSSION TITLE
Reference-based refinement failed

### DISCUSSION THREAD

**Discussion by kexinliu0001** [nextpyp/.github#28] [Q&A] [answered]
Hello, I am running the sample data (EMPIAR-10164) using standalone mode according to the tutorial. The previous steps worked fine, but the Reference-based refinement failed. I'm not sure what the problem is and how to deal with it. and I cannot find /opt/pyp/bin/run/pyp where the error seemed to be.

My setting is the same as shown in tutorial. the stream and logs is:

<img width="1062" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/e7faf208-0d84-4d4e-90a3-106d86b2eff8">
<img width="1081" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/b0edd45d-1e5d-4376-8fe0-be8b0a9a9358">
<img width="1045" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/43d702b7-8654-45e4-b3db-14d9c70299ba">
<img width="1055" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/ab625e28-30f4-4c26-ac44-bfdf1c428ffc">

Moreover, I found that the requests for input parameter file in refinement tab is different from that in tutorial :
<img width="1069" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/bb52592b-2ec2-4f0c-9fca-559226005097">
it asked for  (*.par.bz2), and I found one in the folder where I downloaded the data for the tutorial
<img width="1086" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/4b25efe6-41a9-4e1d-a310-ce24d2fdf5cd">

however, in tutorial the input parameter file should be (.txt)
<img width="820" alt="image" src="https://github.com/nextpyp/.github/assets/110724494/782c98db-9aed-4e5d-a4db-09e28c64856a">

Does the error above result from this difference?
if so, how can I select the text file from "pre-processing"?

**Comment by tofushan** [comment]
Sorry, the description is confusing. It is asking you to provide `*.txt` for the reference-based refinement instead of `*.par.bz2` like other refinement blocks. We're making the distinction for the next release. 

To select the txt file from upstream pre-processing block, you can first clear the selection by clicking the trash can icon, and then click the magnifier. It should direct you to the directory where the txt file is stored. If you still cannot find it, you can get the directory of the pre-processing block by selecting "Show filesystem location" of the block, and it should under `{path_pre-processing}/frealign` folder.

**Reply by kexinliu0001** [reply]
thank you for your reply!

I've resubmitted the job, and it is shown to be running.
![image](https://github.com/nextpyp/.github/assets/110724494/b1c3f21d-3d91-42f6-96e7-c952374408e7)

however, 10 hours passed and the stream and log did not update. is it normal? it seemed to be stuck.
![image](https://github.com/nextpyp/.github/assets/110724494/8ec30af5-29c3-4423-9e23-8a3a6871300d)

![image](https://github.com/nextpyp/.github/assets/110724494/a36f5ed2-8b68-4a87-98b5-9419ff8e61d8)

as I found there are 2 text files in the preprocessing folder:
![image](https://github.com/nextpyp/.github/assets/110724494/c0003734-e4ff-4acd-9567-ea79b745a561)
is it ok to use just the second one when submitting the job?

**Reply by tofushan** [reply, marked as answer]
Regarding the file, you can always use `tomo-preprocessing-xxxxx_volumes.txt`, which is essentially a symbolic link pointing to the file with timestamps (the first one in your screenshot). So in your case, those two files are the same thing. If you re-run the pre-processing, a new txt file will be produced, and the link will also be updated to point to the most up-to-date txt file. 

According to the log, it looks like refinement is running correctly. Unfortunately, as reference-based refinement is computationally expensive, the total running time is extremely long if using standalone mode. You may find similar discussion #23 helpful. We strongly suggest using SLURM in nextPYP for distributed processing if your computing environment allows you to do so.

**Reply by xingzhe1129** [reply]
Hi, I ran particle refienment for EMPIAR-10164 data set with 5 nodes and 4 array jobs ran successfully but always the first array job failed at csp_local_merge step. I repeated it above 5 times but the result was always the same. Do you meet this before? Do you know how to fix it? Thanks.
<img width="264" alt="Screen Shot 2024-01-10 at 23 09 27" src="https://github.com/nextpyp/.github/assets/76068275/9792e69b-2dfb-4d97-ae45-02dce4337615">
![Screen Shot 2024-01-10 at 23 10 51](https://github.com/nextpyp/.github/assets/76068275/b5fec20e-5f91-4e90-b106-4a09aa068811)

**Reply by yezhou867** [reply]
Was the first job always allocated to the same node or a random node?  What's your slurm settings? The first job is always the one with the most particles which may need more memory than others. The error shows the output of the local merge job is missing.  Can't tell too much whether it's because of memory or not.

**Reply by xingzhe1129** [reply]
Thank you so much for your quick response. The first job was not always allocated to the same node. The first job was always running the same TS_01 tomogram. The errors of the first job was not the same each time.
![Screen Shot 2024-01-13 at 19 24 24](https://github.com/nextpyp/.github/assets/76068275/ee7176bc-efa8-4f90-acf3-81d6bbb693cd)
![Screen Shot 2024-01-13 at 19 25 56](https://github.com/nextpyp/.github/assets/76068275/47f59a16-9d7f-476c-9ac5-578fa82688b6)
![Screen Shot 2024-01-13 at 19 27 06](https://github.com/nextpyp/.github/assets/76068275/fcf5cabd-1296-4b08-a6e8-84f8379d8ab8)

**Reply by yezhou867** [reply]
Do you have any information about the scratch space on each node? Looks like for the first jobs, the space could run out at a certain step.

**Reply by yezhou867** [reply]
Could you post the full log for one of the failed jobs,  especially including the first several lines of info stating the machine's basic settings?

**Reply by xingzhe1129** [reply]
Sure. Could you please take a look at the first several lines of the log file?

![Screen Shot 2024-01-13 at 20 55 51](https://github.com/nextpyp/.github/assets/76068275/8b097bea-0efa-46f3-8bf3-ae825d6c25ea)

**Reply by yezhou867** [reply]
The screenshot is missing an important line with the info on how many particles were extracted, just one line below the edge.  Please add the log to where you get the error.  Your available space showing here is 42G, so make sure that's enough for the particles and other intermediate files to save properly.

**Reply by yezhou867** [reply]
The most likely reason that caused this problem is the free space not enough on the computing node.  If there is nothing you can do about the scratch space size, one solution is to reduce the particle number by increasing the value of  `Minimum distance between spikes (voxels)` in the `virion/spike detection` tab during the preprocessing job.  Since other tilt series jobs are OK, you can try to make the TS_01 match a very similar number of particles as the other jobs.

**Reply by xingzhe1129** [reply]
The first tilt series have the most particles, which has 239,158. We tried to increase the scratch space size but the job can't run. I will try to decrease the particle number of the first tilt series. Thanks.

**Reply by xingzhe1129** [reply]
How can we increase the scratch space size? How can we mount the scratch space on other bigger space except /dev/mapper/centos-root? My jobs always failed because the centos-root file is full, which is only 50G. What scratch space size is enough for two tutorial data sets? I see your scratch is 12T, which is so big. Thanks.

**Reply by yezhou867** [reply]
Did you try to configure the instance as described here?
https://nextpyp.app/files/pyp/latest/docs/reference/config.html

**Comment by xingzhe1129** [comment]
Now our nextPYP temporary data was stored in /temp_work/cryosparc-wu , which is 5T and most of the steps before merging are running well but there is always a warning, which said "WARNING: While bind mounting '/temp_work/cryosparc-wu:/temp_work/cryosparc-wu': destination is already in the mount point list". Finally, the job failed, I am not sure whether the warning is the reason of job failure. Could you please take a look at the attachments and figure out the failure reason? Thank you so much. 
![Screen Shot 2024-01-18 at 20 23 36](https://github.com/nextpyp/.github/assets/76068275/c0663259-147e-4f74-9b5e-846c607bca01)
![Screen Shot 2024-01-19 at 12 14 49](https://github.com/nextpyp/.github/assets/76068275/d257981d-e42e-4aff-9f18-6032b61e5966)
![Screen Shot 2024-01-19 at 12 15 14](https://github.com/nextpyp/.github/assets/76068275/e9065241-7808-47ad-8802-6ac14575441b)
![Screen Shot 2024-01-19 at 12 15 34](https://github.com/nextpyp/.github/assets/76068275/337afa56-1ff6-49f6-95d0-5325e7d54859)
![Screen Shot 2024-01-19 at 12 15 46](https://github.com/nextpyp/.github/assets/76068275/f11125b2-c30a-4ada-b2c7-f0340ef48abf)

**Reply by yezhou867** [reply]
Could you tell us a little more about the job you were running? Was the dose weighting enabled?   
And please share your config.toml settings here, so we can check what could be wrong.

**Reply by xingzhe1129** [reply]
I am running Region-based local refinement (before masking) for empiar-10164 data set and I remembered that this step doesn't have dose weighting. Please check the attached config. Thanks.

<img width="694" alt="Screen Shot 2024-01-18 at 20 20 10" src="https://github.com/nextpyp/.github/assets/76068275/815d75db-244d-4f53-b334-dfd270092c70">

**Reply by yezhou867** [reply]
Just want to make sure, were you setting the bundle size as 5? Looks like the merge trying to merge 5 files into a single file and messed up the reconstruction logic.   And when you launch the jobs, were the job ids set by slurm the same? It is strange that you have 5 files in the merge list.

**Reply by xingzhe1129** [reply]
The bundles size was 1 and I don’t know why it can’t run parallel. Before we set this new temporary file, it can run parallel

**Reply by yezhou867** [reply]
Probably need to set the slurm configuration in the right way by the root user so that each job will write in the scratch space with a different id name. 
Temporarily, can try the bundle size of 5, so everything runs in a single job properly.

**Reply by xingzhe1129** [reply]
I tried with bundle size of 5 but it failed with the same reason. The error said that AssertionError: Number of intermediate reconstructions does not match (half1: 0 is NOT half2: 0), which is very similar to before with shortage of storage for temporary files. I can run Reference-based refinement and Fully constrained refinement successfully and I don't know why region-based local refinement failed. The attachment is the full config. Could you please take a look and help us figure it out? Thank you so much.

![Screen Shot 2024-01-19 at 17 21 00](https://github.com/nextpyp/.github/assets/76068275/a1ba500d-76e4-42bf-a3ba-0c6738ae81b6)

**Reply by yezhou867** [reply]
The configure file looks fine.  
To understand the situation more clearly.  Can you share more information about the following questions?
1. Did previous jobs also run with this new scratch space?
2. What is the output of the filter job before this region-based refinement? 
3. If you could also provide a log from the successful fully constrained refinement job would be easier for us to compare. Since there should not be too much difference between those 2 jobs.

**Reply by xingzhe1129** [reply]
I tried to skip the filter job and ran region-based refinement successfully and will try to take a close look at the filter job. I like nextPYP and it will be perfect if we can ran all refinements with both GPUs and CPUs. Thanks.

**Comment by JuhaHuiskonen** [comment]
Hi, I am having the same error. 

```
17	Traceback (most recent call last):
18	  File "/opt/pyp/bin/run/pyp", line 3914, in <module>
19	    csp_swarm(args.file, parameters, int(args.iter), args.skip, args.debug)
20	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
21	    return func(*args, **kwargs)
22	  File "/opt/pyp/bin/run/pyp", line 2376, in csp_swarm
23	    [allboxes, allparxs] = csp_extract_coordinates(
24	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
25	    return func(*args, **kwargs)
26	  File "/opt/pyp/src/pyp/inout/metadata/core.py", line 1869, in csp_extract_coordinates
27	    [allboxes, allparxs] = tomo_extract_coordinates(
28	  File "/opt/pyp/src/pyp/inout/metadata/core.py", line 2007, in tomo_extract_coordinates
29	    refinement = project_params.resolve_path(parameters["refine_parfile_tomo"])
30	KeyError: 'refine_parfile_tomo'
31	2024-02-08 12:09:38 [ERROR] /opt/pyp/bin/run/pyp:3927 | PYP (cspswarm) failed
32	2024-02-08 12:09:42 [INFO] /opt/pyp/bin/run/pyp:3273 | Job 3185360_1 (v0.6.2) launching on c1118.mahti.csc.fi using 128 task(s) and 200 GB of RAM
33	2024-02-08 12:09:42 [INFO] pyp/refine/csp/particle_cspt.py:57 | Total time elapsed (csp_local_merge): 00h 00m 00s
34	Traceback (most recent call last):
35	  File "/opt/pyp/bin/run/pyp", line 3992, in <module>
36	    particle_cspt.merge_movie_files_in_job_arr(
37	  File "/opt/pyp/src/pyp/utils/timer.py", line 78, in wrapper_timer
38	    return func(*args, **kwargs)
39	  File "/opt/pyp/src/pyp/refine/csp/particle_cspt.py", line 380, in merge_movie_files_in_job_arr
40	    with open(movie_file) as f:
41	FileNotFoundError: [Errno 2] No such file or directory: 'stacks.txt'
42	2024-02-08 12:09:42 [ERROR] /opt/pyp/bin/run/pyp:4012 | PYP (csp_local_merge) failed
43	2024-02-08 12:09:43 [INFO] /opt/pyp/bin/run/pyp:4018 | Deleted temporary files from /scratch/project_2009485/nextPYP/tmp/huiskone/3185360_1
```

I have given the `original_volumes.txt` file instead of the `par.bz2` file, as follows

<img width="1499" alt="Screenshot 2024-02-08 at 13 00 44" src="https://github.com/nextpyp/.github/assets/20352352/bfdd5a0b-b4cc-4c0c-97f4-915b38d14d98">

Any suggestions how to proceed?  I have tried rerunning the preprocessing.

**Reply by yezhou867** [reply]
https://github.com/orgs/nextpyp/discussions/57#discussioncomment-8361155

Should be the other way to specify the input parameters.
